src/neml2/models/Model.cxx

```diff
@@ -33,266 +33,414 @@ Model::expected_options()
 {
   OptionSet options = Data::expected_options();
   options += NonlinearSystem::expected_options();
-  options.set<std::vector<LabeledAxisAccessor>>("additional_outputs");
+  options.set<std::vector<VariableName>>("additional_outputs");
   options.set<bool>("use_AD_first_derivative") = false;
   options.set<bool>("use_AD_second_derivative") = false;
+  options.set<int>("_extra_derivative_order") = 0;
   return options;
 }
 
 Model::Model(const OptionSet & options)
   : Data(options),
-    ParameterStore(options),
+    ParameterStore(options, this),
+    VariableStore(options, this),
     NonlinearSystem(options),
-    _input(declare_axis()),
-    _output(declare_axis()),
+    _additional_outputs(options.get<std::vector<VariableName>>("additional_outputs")),
     _AD_1st_deriv(options.get<bool>("use_AD_first_derivative")),
-    _AD_2nd_deriv(options.get<bool>("use_AD_second_derivative"))
+    _AD_2nd_deriv(options.get<bool>("use_AD_second_derivative")),
+    _options(default_tensor_options()),
+    _deriv_order(-1),
+    _extra_deriv_order(options.get<int>("_extra_derivative_order"))
 {
   check_AD_limitation();
-  for (const auto & var : options.get<std::vector<LabeledAxisAccessor>>("additional_outputs"))
-    _additional_outputs.insert(var);
-  setup();
 }
 
 void
-Model::to(const torch::TensorOptions & options)
+Model::setup()
 {
-  // This takes care of all the buffers recursively
-  Data::to(options);
+  // Declare nonlinear parameters as additional input variables
+  for (const auto & [name, param] : nl_params())
+    declare_input_variable(param->base_storage(), VariableName(param->name()));
 
-  send_parameters_to(options);
-  for (auto & model : _registered_models)
-    model->to(options);
+  // Setup input and output axes
+  setup_layout();
+
+  // Setup functional dependence for each output variable
+  for (auto && [y_name, y_var] : output_views())
+  {
+    y_var.clear_args();
+    for (auto && [x_name, x_var] : input_views())
+      y_var.add_arg(x_var);
+  }
+
+  // Determine if this is an implicit system
```

Comment:
Were we doing this before?  This seems really brittle.  Isn't there a better way to set this instead of checking based on the model inputs.

Reply:
We were manually setting this parameter before. In what way this might be brittle?